sql/sql_optimizer.cc

```diff
@@ -137,371 +137,371 @@ static uint32 get_key_length_tmp_table(Item *item);
-  ASSERT_BEST_REF_IN_JOIN_ORDER(this);
-
-  if (zero_result_cause)
-    goto setup_subq_exit;
+    if ((where_cond || group_list || order) &&
+        substitute_gc(thd, select_lex, where_cond, group_list, order))
+    {
+      // We added hidden fields to the all_fields list, count them.
+      count_field_types(select_lex, &tmp_table_param, select_lex->all_fields,
+                        false, false);
+    }
 
-  if (rollup.state != ROLLUP::STATE_NONE)
-  {
-    if (rollup_process_const_fields())
+    // Set up join order and initial access paths
+    THD_STAGE_INFO(thd, stage_statistics);
+    if (make_join_plan())
     {
-      DBUG_PRINT("error", ("Error: rollup_process_fields() failed"));
+      if (thd->killed)
+        thd->send_kill_message();
+      DBUG_PRINT("error",("Error: JOIN::make_join_plan() failed"));
       DBUG_RETURN(1);
     }
-    /*
-      Fields may have been replaced by Item_func_rollup_const, so
-      recalculate the number of fields and functions for this query block.
-    */
 
-    // JOIN::optimize_rollup() may set quick_group=0, and we must not undo that.
-    const uint save_quick_group= tmp_table_param.quick_group;
+    // At this stage, join_tab==NULL, JOIN_TABs are listed in order by best_ref.
+    ASSERT_BEST_REF_IN_JOIN_ORDER(this);
 
-    count_field_types(select_lex, &tmp_table_param, all_fields, false, false);
-    tmp_table_param.quick_group= save_quick_group;
-  }
-  else
-  {
-    /* Remove distinct if only const tables */
-    select_distinct&= !plan_is_const();
-  }
+    if (zero_result_cause)
+      goto setup_subq_exit;
 
-  if (const_tables && !thd->locked_tables_mode &&
-      !(select_lex->active_options() & SELECT_NO_UNLOCK )&& part!=3) //TIANMU UPGRADE
-  {
-    TABLE *ct[MAX_TABLES];
-    for (uint i= 0; i < const_tables; i++)
-      ct[i]= best_ref[i]->table();
-    mysql_unlock_some_tables(thd, ct, const_tables);
-  }
-  if (!where_cond && select_lex->outer_join)
-  {
-    /* Handle the case where we have an OUTER JOIN without a WHERE */
-    where_cond=new Item_int((longlong) 1,1);	// Always true
-  }
+    if (rollup.state != ROLLUP::STATE_NONE)
+    {
+      if (rollup_process_const_fields())
+      {
+        DBUG_PRINT("error", ("Error: rollup_process_fields() failed"));
+        DBUG_RETURN(1);
+      }
+      /*
+        Fields may have been replaced by Item_func_rollup_const, so
+        recalculate the number of fields and functions for this query block.
+      */
 
-  error= 0;
-  /*
-    Among the equal fields belonging to the same multiple equality
-    choose the one that is to be retrieved first and substitute
-    all references to these in where condition for a reference for
-    the selected field.
-  */
-  if (where_cond)
-  {
-    where_cond= substitute_for_best_equal_field(where_cond, cond_equal,
-                                                map2table);
-    if (thd->is_error())
+      // JOIN::optimize_rollup() may set quick_group=0, and we must not undo that.
```

Comment:
"must undo" or "must not undo"

Reply:
this is MySQL source code: `must not undo`